include/exadg/poisson/spatial_discretization/operator.cpp

```diff
@@ -98,7 +119,8 @@ Operator<dim, Number, n_components>::distribute_dofs()
 
   dof_handler.distribute_dofs(*fe);
 
-  dof_handler.distribute_mg_dofs();
+  if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+    dof_handler.distribute_mg_dofs();
```

Comment:
This is very difficult to understand for non-dealii principal developers. What is the background?

Reply:
`DoFHandler::distribute_mg_dofs()` is not working for simplex meshes. To be able to use multigrid, one needs to use global coarsening for now.

Reply:
Thanks. This implies that we need more asserts for the simplex branch related to which preconditioners can be used (with additional explanations). The reasons behind can not be deduced from the above if statement and the above if would still allow to select „standard dealii multigrid“ for simplex, which would then probably crash at some point.

@kronbichler as mentioned elsewhere in the past (and not directly related to this PR), distribute_mg_dofs() is a problem in terms of design and we need to discuss this in a broader context in my opinion.

Reply:
Regarding `distribute_mg_dofs`: I agree that this is annoying and that it would be good to be able to eliminate it. From an application point of view, one could argue that we already have the ingredients to achieve this, as we might simply switch all solvers to global coarsening. One could achieve similar (or the same) partitioning with that setting, so the parallel scalability of the solver would be the same. Furthermore, one could pull over the optimizations from the local smoothing in terms of data access to global coarsening, eliminating any possible performance penalty still visible for the latter. (I have not measured lately, but maybe @peterrum has.) The setup would remain somewhat more expensive due to the need to create additional triangulations, but for ExaDG this is essentially a one-time cost. Also the increase in memory consumption should be tolerable in general, compared to other data structures we are using. (The scalar Poisson and convection-diffusion modules might be the only really seeing a noticeable effect.)

My main concern at this stage is how well it is possible to hide global coarsening data structures in a complex application scenario: What about the case where multiple fields want to create a multigrid hierarchy. Can we collect all information in the `Grid` class or, more generally, in a central data structure, such that we avoid going through these steps more than once? This is because my reasoning above regarding costs relies on the fact that we might not do this more than once per grid. Think about dynamic adaptivity where we change the mesh more often.

Opinions?

Reply:
For the standard dealii multigrid approach (I mean local smoothing), can we be more conservative in the sense that `distribute_mg_dofs()` is only done in multigrid and that some data structures need to be set up twice in the worst case? What about the performance penalty for this case?

Reply:
> Can we collect all information in the Grid class ...?

The question is whether we then simply shift the problem? We would then have multigrid functionality/data structures in a class (`Grid`) that is used very generally and not just for multigrid.

Reply:
Thinking about this, I see two alternatives: Either we call `dof_handler.distribute_mg_dofs` inside the multigrid solver (possibly doing `const_cast<DoFHandler&>` on the object, to only call it on demand and at the place where we need it. The other alternative is to set up an additional `DoFHandler` on which we can call `distribute_mg_dofs`, keeping everything isolated to the multigrid solver. For the latter, the perspective of `DoFHandler` has the potential problem that we might have a renumbering on the original `DoFHandler`, for which I do not see a simple way right now to make sure they match. We do not do any renumbering at this point, but we might possibly do in the future and should keep responsibilities clear.

From a performance perspective in the solver classes, there would not be any penalty in this case against what we have now. The second option would pay a bit in terms of storage cost of the additional `DoFHandler`.

Reply:
I would prefer such an approach. Our multigrid implementation has the option `update()`, which updates the multigrid preconditioner in case something has changed (e.g. parameters, mesh motion, etc.). Maybe the problem that you mention would fit into this logic.

Reply:
> >  Can we collect all information in the Grid class ...?
> 
> The question is whether we then simply shift the problem? We would then have multigrid functionality/data structures in a class (Grid) that is used very generally and not just for multigrid.

I agree, it does put a burden on that class and leads to questions about the responsibility. However, this is in conflict with the desire to expose information created for a particular linear solver (regarding the sharing of say a full multigrid hierarchy created from a single triangulation) to another solver. Hence, I see that we need to decide how to balance the conflict between the level of separation and performance/memory costs associated to repeatedly do the same thing. In consequence, we would need to decide who would keep track of such "hidden states" going from one solver to the other.

Reply:
> I would prefer such an approach.

Let me have a look at it later today or on Monday, when I've finished some other topics.

Reply:
> I agree, it does put a burden on that class and leads to questions about the responsibility. However, this is in conflict with the desire to expose information created for a particular linear solver (regarding the sharing of say a full multigrid hierarchy created from a single triangulation) to another solver. Hence, I see that we need to decide how to balance the conflict between the level of separation and performance/memory costs associated to repeatedly do the same thing. In consequence, we would need to decide who would keep track of such "hidden states" going from one solver to the other.

It would be acceptable in the grid class if it really works for all cases, but not if we would need if-else-branches depending on other parameters of ExaDG. I don't have the overview of which type of coarsening works for which types of dealii-triagnulations and which types of elements ..

Reply:
Regarding `distribute_mg_dofs()`: @kronbichler, the first option you described previously

> Either we call dof_handler.distribute_mg_dofs inside the multigrid solver (possibly doing const_cast<DoFHandler&> on the object), to only call it on demand and at the place where we need it.

seems to be most straight-forward to realize. Could you make a suggestion? I could also help you. However, I now want to realize the simplex-support for Poisson and I currently ask myself how to best solve the `distribute_mg_dofs()` problem (so it would be helpful if we could make progress here).

Reply:
After `distribute_mg_dofs()` is called, the code deals with affine constraints (`DoFTools::make_zero_boundary_constraints()`). In order to shift `distribute_mg_dofs()` to the multigrid preconditioner (which is set up later in the code), we have to make sure that other parts of the code (such as the boundary conditions mentioned above) do not rely on the call `distribute_mg_dofs()`.